### ISSUE TITLE
kubemark is still failing.

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#41713]
The PR https://github.com/kubernetes/kubernetes/pull/41703 fixed hollow npd. Now hollow npd can work properly. However, I found that npd can't update node condition, and keep reporting error:
```
E0219 10:05:23.100125       1 manager.go:160] failed to update node conditions: nodes "hollow-node-w9fkf" not found
```
The reason is that hollow kubelet never registers the node.

I checked the hollow kubelet. Most of them are empty, but in some of them I saw tons of the following log:
```
0219 09:59:54.303624       9 file.go:72] unable to read config path "/tmp/hollow-kubelet.037396831/manifest013003570": unable to create i
notify: inotify_init: too many open files
```

BTW, we really need hollow node log. I can't reproduce this in my own kubemark cluster, I can only wait for the kubemark test running, and login into the running node to get some debug information.

/cc @shyamjvs @wojtek-t @gmarek

**Comment by Random-Liu** [comment]
I think the reason is that each npd will also create an `inotify` instance now.

On the node, the inotify instance limit is:
```
fs.inotify.max_queued_events = 16384
fs.inotify.max_user_instances = 128
fs.inotify.max_user_watches = 8192
```
Previously, we have **500 hollow nodes, 6 real nodes**. It means that on each node there are **83-84 inotify instances** held by hollow kubelet.

However, now each hollow npd will also hold an inotify instance. Then we have **166-168 inotify instances** on each node, which exceeds the `128` limit.

There are 3 options to fix this:
* Option 1: Increase the real node number in kubemark.
* Option 2: Use sysctl to change the inotify instance limit on each node.
* Option 3: Change the NPD behavior, add a completely idle mode.

Even though we fix NPD for now, in the future we could still hit this when any node component uses one more inotify...

/cc @kubernetes/sig-node-bugs @kubernetes/sig-scalability-misc

**Comment by Random-Liu** [comment]
@foxish I'm fine with reverting the PR https://github.com/kubernetes/kubernetes/pull/41703 and https://github.com/kubernetes/kubernetes/pull/40206 to unblock the queue.

Will send a fix for the new issue...https://github.com/kubernetes/kubernetes/issues/41713#issuecomment-280911500

**Comment by foxish** [comment]
The kubemark test was passing briefly I think, after the earlier fix. And it seems like something merged 2 hours ago. Could one of those be related?

**Comment by Random-Liu** [comment]
@foxish Nope. There are 2 issues here. The first one got resolved by my PR #41703, however this one could not be resolved very soon.

If I go with option 3, I'll need to update NPD and cut a new release.

The option 1, 2 need the opinion from Warsaw team.

Warsaw team doesn't have president day. If we could wait till their Monday morning (our Sunday night), they may have better opinion about this. I don't know whether we could wait till then. :P

**Comment by foxish** [comment]
Revert makes sense I think. We can deal with the larger problem later on during the week I imagine, since it's not a decision that seems obvious. I created https://github.com/kubernetes/kubernetes/pull/41715 reverting https://github.com/kubernetes/kubernetes/pull/41703

Looks like https://github.com/kubernetes/kubernetes/pull/40206 needs manually creating the revert PR. Can you do that when you find the time?

**Comment by Random-Liu** [comment]
Discussed with @foxish. We decided to revert #41703 and partially revert #40206 to only revert changes in `test/kubemark`. Because new npd is actually working well in the real cluster. The only problem is this issue.

Only revert kubemark related change can unblock the submit queue, and will also give npd more soak time.

**Comment by gmarek** [comment]
Thanks for unblocking the queue!

**Comment by wojtek-t** [comment]
Thanks - I think it was a good decision.

I definitely think that option 3 is the worst one - I think we want all component to behave as similar to to prod environment as possible.

Between 1 and 2, I think (2) is better, though I'm not sure how big changes this would be. If this is relatively big change, then (1) might be easier to do.

@gmarek @shyamjvs

**Comment by shyamjvs** [comment]
+1 to @wojtek-t 's comment.
It shouldn't really be difficult to change the start-kubemark.sh script to ssh to each of the nodes and run sysctl setting inotify limit to higher value (though I'm not sure how long this would take when we have higher no. of nodes, like 60 in case of kubemark-scale test). We should ideally be doing this in parallel or set the var during the real node creation itself, if possible. Increasing real node no. in kubemark just because of the inotify limit doesn't seem like the best thing to do.

**Comment by shyamjvs** [comment]
@Random-Liu If we agree on the 2nd option, I can change the code for it. Also FYI, if you need hollow-node logs to debug stuff, they should be available once PR https://github.com/kubernetes/kubernetes/pull/41739 gets merged.

**Comment by Random-Liu** [comment]
> @Random-Liu If we agree on the 2nd option, I can change the code for it. Also FYI, if you need hollow-node logs to debug stuff, they should be available once PR #41739 gets merged.

Thanks a lot for the hollow-node logs!

Going with option 2 sounds good to me. :)

**Comment by gmarek** [comment]
It may be a stupid question, but is it possible to build a container image that, when run in a privileged mode will bump ionotify limit and then run it similarly to image-puller?

**Comment by wojtek-t** [comment]
> It may be a stupid question, but is it possible to build a container image that, when run in a privileged mode will bump ionotify limit and then run it similarly to image-puller?

Something like that was also on my mind when I wrote that it shouldn't be that difficult.

I don't want to put things like "ssh to all nodes and do something" to our startup scripts.

**Comment by yujuhong** [comment]
> It may be a stupid question, but is it possible to build a container image that, when run in a privileged mode will bump ionotify limit and then run it similarly to image-puller?

You don't need to build a container image. Just need to create a pod using some generic image (e.g., busybox) and pass the command (`sysctl ...`) in the pod spec.

**Comment by shyamjvs** [comment]
So based on discussion with @Random-Liu , we seem to have the following options to achieve what we want:
1. Add a daemonset on each node and run sysctl (this seems to be an overkill to me as we just want to run sysctl command)
2. Add a privileged pod on each node for running sysctl (but the pod needs to be passed the node name for it and also the logic doesn't seem trivial to me)
3. SSH to each node and run sysctl on it (but @wojtek-t doesn't seem to be inclined towards this idea)
4. Run sysctl command in each hollow node using an init container (this would run sysctl multiple times redundantly on the same node, but it's my personal favourite)

For all the above options, we might need a logic for computing the new inotify limit based on no. of hollow-nodes and no. of real nodes (something like `inotify_limit = min(default_inotify_limit, (avg no. of hollow-nodes per node) * 2 + some_extra_value )`)

**Comment by wojtek-t** [comment]
> Run sysctl command in each hollow node using an init container (this would run sysctl multiple times redundantly on the same node, but it's my personal favourite)

I'm fine with that. I don't think running sysctl multiple times is a problem (all but first call should be no-ops, right?)

> For all the above options, we might need a logic for computing the new inotify limit based on no. of hollow-nodes and no. of real nodes (something like inotify_limit = min(default_inotify_limit, (avg no. of hollow-nodes per node) * 2 + some_extra_value)))

That's overkill. Let's assume we can't have more than 100 hollow-nodes on a given node and set it to 250 or sth like that (I mean a constant number).

**Comment by shyamjvs** [comment]
sgtm

**Comment by Random-Liu** [comment]
@shyamjvs Option 3 sounds good to me. Thanks for working on this!
